app/javascript/packages/document-capture/components/full-screen.jsx

```diff
@@ -1,81 +1,139 @@
-  const trapRef = useRef(/** @type {?import('focus-trap').FocusTrap} */ (null));
+function useFocusTrap(containerRef, onRequestClose) {
+  const trapRef = useRef(/** @type {FocusTrap?} */ (null));
   const onRequestCloseRef = useRef(onRequestClose);
+
   useEffect(() => {
     // Since the focus trap is only initialized once, but the callback could
     // be changed, ensure that the current reference is kept as a mutable value
     // to reference in the deactivation.
     onRequestCloseRef.current = onRequestClose;
   }, [onRequestClose]);
 
-  const setFocusTrapRef = useCallback((node) => {
-    if (trapRef.current) {
-      trapRef.current.deactivate();
-    }
-
-    if (node) {
-      trapRef.current = createFocusTrap(node, {
+  useEffect(() => {
+    if (containerRef.current) {
+      trapRef.current = createFocusTrap(containerRef.current, {
         onDeactivate: () => onRequestCloseRef.current(),
         clickOutsideDeactivates: true,
       });
 
       trapRef.current.activate();
     }
-  }, []);
-
-  useEffect(() => {
-    if (activeInstances++ === 0) {
-      document.body.classList.add('has-full-screen-overlay');
-      document.documentElement.classList.add('has-full-screen-overlay');
-    }
 
     return () => {
-      if (--activeInstances === 0) {
-        document.body.classList.remove('has-full-screen-overlay');
-        document.documentElement.classList.remove('has-full-screen-overlay');
-      }
+      trapRef.current?.deactivate();
     };
   }, []);
+}
+
+const useBodyClass = (() => {
+  /**
+   * @type {WeakMap<any, number>}
+   */
+  const activeInstancesByComponent = new WeakMap();
 
   return (
-    <div ref={setFocusTrapRef} aria-modal="true" className="full-screen bg-white">
+    /**
+     * @param {string} className Class name to add to body element
+     * @param {React.ComponentType<any>} Component React component definition
+     */
+    (className, Component) => {
+      /**
+       * Increments the number of active instances for the current component by the given amount,
+       * adding or removing the body class for the first and last instance respectively.
+       *
+       * @param {number} amount
+       */
+      function incrementActiveInstances(amount) {
+        const activeInstances = activeInstancesByComponent.get(Component) || 0;
+        const nextActiveInstances = activeInstances + amount;
+
+        if (nextActiveInstances === 1 || nextActiveInstances === 0) {
+          document.body.classList.toggle(className, nextActiveInstances > 0);
+          document.documentElement.classList.toggle(className, nextActiveInstances > 0);
+        }
+
+        activeInstancesByComponent.set(Component, nextActiveInstances);
+      }
+
+      useEffect(() => {
+        incrementActiveInstances(1);
+        return () => incrementActiveInstances(-1);
+      }, []);
+    }
+  );
+})();
+
+/**
+ * @param {React.MutableRefObject<HTMLElement?>} containerRef
+ */
+function useSoleAccessibleContent(containerRef) {
```

Comment:
another name idea, what about something like `ariaHideSiblingElements` something?

Reply:
> another name idea, what about something like `ariaHideSiblingElements` something?

I struggled with this one as well. I also like the idea of incorporating "siblings" into the name. I thought about using the term [`inert`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/inert) (e.g. `useInertSiblingElements`), since the purpose of `inert` is effectively what we're trying to implement here, except at this hook we're only dealing with AT focus and not keyboard focus.

Alternatively, it would make more sense to call it as such if we pulled in an [`inert` polyfill](https://github.com/WICG/inert) to use it in place of what we're implementing here as `useSoleAccessibleContent` and `useFocusTrap`.

Reply:
> Alternatively, it would make more sense to call it as such if we pulled in an [`inert` polyfill](https://github.com/WICG/inert) to use it in place of what we're implementing here as `useSoleAccessibleContent` and `useFocusTrap`.

While I think the attribute is well-suited for this use-case, the caveats described in ["Legacy Browser Support"](https://github.com/WICG/inert#legacy-browser-support) and ["Performance and gotchas"](https://github.com/WICG/inert#performance-and-gotchas) give me some pause, and it's [not clear](https://chromestatus.com/feature/5703266176335872) if or when any browser will actually implement it. So I might put this idea on ice for now.

Reply:
Also, to this and your previous comment, since these are designed to be used as hooks, I'll at least want to keep the `use` prefix in both cases.

>A custom Hook is a JavaScript function whose name starts with ”use” and that may call other Hooks.

https://reactjs.org/docs/hooks-custom.html#extracting-a-custom-hook

Reply:
Went with `useInertSiblingElements` in 44029b1dbab8538998e992b69c1b98cb89cc1df1. LMK if there's still room for improvement. Per my previous comment, _technically_ "inert" means more than just `aria-hidden`, but I'm pretty okay with it, especially since I didn't extract it for reuse.